### ISSUE TITLE
Slow custom loss function

### ISSUE THREAD

**Issue by Moelf** [dmlc/XGBoost.jl#213]
```julia
using DifferentiationInterface, ForwardDiff

function my_loss(yhat, y)
    CrossEntropyLoss()(yhat, y)
end
my_loss_d(ŷ, y) = derivative(ζ -> my_loss(ζ, y), AutoForwardDiff(), ŷ)
my_loss_dd(ŷ, y) = derivative(ζ -> my_loss_d(ζ, y), AutoForwardDiff(), ŷ)


xgboost(train_data, my_loss_d, my_loss_dd; tree_method="gpu_hist", objective="binary:logistic", num_round=10)
```

This is much slower than if not providing my own loss -- I suspect this is because these functions are scalar version and we also create an anonymous function for every value?

**Comment by Moelf** [comment]
https://github.com/dmlc/XGBoost.jl/blob/361022aba088978719f3d1dd5cb793882e833825/src/booster.jl#L387-L388

eh, is it because now it's training on CPU?
